include/boost/geometry/algorithms/detail/distance/linear_or_areal_to_areal.hpp

```diff
@@ -27,65 +27,63 @@ namespace detail { namespace distance
 {
 
 
-template <typename Linear, typename Areal, typename Strategy>
+template <typename Linear, typename Areal, typename Strategies>
 struct linear_to_areal
 {
     typedef typename strategy::distance::services::return_type
         <
-            Strategy,
+            decltype(std::declval<Strategies>().distance(std::declval<Linear>(), std::declval<Areal>())),
             typename point_type<Linear>::type,
             typename point_type<Areal>::type
         >::type return_type;
 
     static inline return_type apply(Linear const& linear,
                                     Areal const& areal,
-                                    Strategy const& strategy)
+                                    Strategies const& strategies)
     {
-        if ( geometry::intersects(linear, areal,
-                                  strategy.get_relate_segment_segment_strategy()) )
+        if ( geometry::intersects(linear, areal, strategies) )
         {
             return 0;
         }
 
         return linear_to_linear
             <
-                Linear, Areal, Strategy
-            >::apply(linear, areal, strategy, false);
+                Linear, Areal, Strategies
+            >::apply(linear, areal, strategies, false);
     }
 
 
     static inline return_type apply(Areal const& areal,
                                     Linear const& linear,
-                                    Strategy const& strategy)
+                                    Strategies const& strategies)
     {
-        return apply(linear, areal, strategy);
+        return apply(linear, areal, strategies);
     }
 };
 
-template <typename Areal1, typename Areal2, typename Strategy>
+template <typename Areal1, typename Areal2, typename Strategies>
 struct areal_to_areal
 {
     typedef typename strategy::distance::services::return_type
         <
-            Strategy,
+            decltype(std::declval<Strategies>().distance(std::declval<Areal1>(), std::declval<Areal2>())),
             typename point_type<Areal1>::type,
             typename point_type<Areal2>::type
         >::type return_type;
 
     static inline return_type apply(Areal1 const& areal1,
```

Comment:
The rules for C++14 `auto` type deduction requires that in all places the same type is returned. Here in one place it's the return type of `linear_to_linear<...>::apply(...)` and in the other place it's `int` (`0`). I could use `?:` operator instead of `if` but then we'd be dealing with the rules defining the type of this expression. In particular with the rule saying that if both return types are convertible to each other then the program is ill-formed. So e.g. this would not compile for a user-defined numeric type constructible from and convertible to `int`.

In short, I'd still have to define some return type inside this function and convert `0` to it before returning it. And I think it's not worth it.

What I could do instead would be using `bg::distance_result<Areal1, Areal2, Strategies>::type` or implement a simple template alias to avoid using heavier `distance_result`:
```
template <typename G1, typename G2, typename Strategies>
using distance_return_t = typename strategy::distance::services::return_type
    <
        decltype(std::declval<Strategies>().distance(std::declval<G1>(), std::declval<G2>())),
        typename point_type<G1>::type,
        typename point_type<G2>::type
    >::type;
```
then the lines above would look like this:
```
typedef distance_return_t<Areal1, Areal2, Strategies> return_type;
```
Unless you have a different idea?

Reply:
>In short, I'd still have to define some return type inside this function and convert 0 to it before returning it. And I think it's not worth it.

Agree, thanks for the explanations. 

> Unless you have a different idea?

The alias make sense since this is used (as far as I remember) in a few places in distance implementation. But I am also OK leaving it as is, since the use of `auto` is problematic.

Reply:
I refactored the return types and some other typedefs. For this purpose I added several template aliases like the one above. Please see if you're ok with the changes.